test/kotlin/integration/ReceiveDataTest.kt

```diff
@@ -0,0 +1,110 @@
+import org.assertj.core.api.Assertions.fail
+import org.junit.Test
+
+private const val apiListenerType =
+  "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager"
+private const val assertionFilterType = "type.googleapis.com/envoymobile.extensions.filters.http.assertion.Assertion"
+private const val assertionResponseBody = "response_body"
+private const val config =
+  """
+    static_resources:
+      listeners:
+      - name: base_api_listener
+        address:
+          socket_address:
+            protocol: TCP
+            address: 0.0.0.0
+            port_value: 10000
+        api_listener:
+          api_listener:
+            "@type": $apiListenerType
+            stat_prefix: hcm
+            route_config:
+              name: api_router
+              virtual_hosts:
+                - name: api
+                  domains:
+                    - "*"
+                  routes:
+                    - match:
+                        prefix: "/"
+                      direct_response:
+                        status: 200
+                        body:
+                          inline_string: $assertionResponseBody
+            http_filters:
+              - name: envoy.filters.http.assertion
+                typed_config:
+                  "@type": $assertionFilterType
+                  match_config:
+                    http_request_headers_match:
+                      headers:
+                        - name: ":authority"
+                          exact_match: example.com
+              - name: envoy.router
+                typed_config:
+                  "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+    """
+
+class ReceiveDataTest {
+
+  init {
+    JniLibrary.loadTestLibrary()
+  }
+
+  @Test
+  fun `response headers and response data call onResponseHeaders and onResponseData`() {
+
+    val engine = EngineBuilder(Custom(config)).build()
+    val client = engine.streamClient()
+
+    val requestHeaders = RequestHeadersBuilder(
+      method = RequestMethod.GET,
+      scheme = "https",
+      authority = "example.com",
+      path = "/test"
+    )
+      .addUpstreamHttpProtocol(UpstreamHttpProtocol.HTTP2)
+      .build()
+
+    val headersExpectation = CountDownLatch(1)
+    val dataExpectation = CountDownLatch(1)
+
+    var status: Int? = null
+    var body: ByteBuffer? = null
+    client.newStreamPrototype()
+      .setOnResponseHeaders { responseHeaders, _ ->
+        status = responseHeaders.httpStatus
+        headersExpectation.countDown()
+      }
+      .setOnResponseData { data, _ ->
+        body = data
+        dataExpectation.countDown()
+      }
+      .setOnError { fail("Unexpected error") }
+      .start()
+      .sendHeaders(requestHeaders, true)
+
+    headersExpectation.await(10, TimeUnit.SECONDS)
+    dataExpectation.await(10, TimeUnit.SECONDS)
+    engine.terminate()
+
+    assertThat(headersExpectation.count).isEqualTo(0)
```

Comment:
Interesting. You like to assert inline in the test flow. You've mentioned this before.

Reply:
I can change it...

Reply:
Except the expectation counts because we'll want to make sure we do an await + check at the end just in case the latches don't get counted down

Reply:
Whoops I meant to put this comment below on your assertions on headers and data.